File: iris-crypto/iris-crypto-rsa/src/main/java/city/smartb/iris/crypto/rsa/RSAKeyPairDecoderBase64.java
package city.smartb.iris.crypto.rsa;

import org.bouncycastle.util.io.pem.PemObject;

import java.security.*;
import java.security.interfaces.RSAPrivateKey;
import java.security.interfaces.RSAPublicKey;
import java.security.spec.PKCS8EncodedKeySpec;
import java.security.spec.X509EncodedKeySpec;
import java.util.Base64;

public class RSAKeyPairDecoderBase64 {

    public static KeyPair decode(String publicKey, String privateKey) throws InvalidKeyException {
        PublicKey pub = decodePublicKey(publicKey);
        PrivateKey priv = decodePrivateKey(privateKey);
        return new KeyPair(pub, priv);
    }

    public static RSAPrivateKey decodePrivateKey(String value) throws InvalidKeyException {
        try {
            KeyFactory kf = KeyFactory.getInstance("RSA");
            PKCS8EncodedKeySpec keySpecPKCS8 = new PKCS8EncodedKeySpec(decode(value));
            return (RSAPrivateKey) kf.generatePrivate(keySpecPKCS8);
        } catch (Exception e) {
            throw new InvalidKeyException("Invalid private public key", e);
        }
    }

    public static RSAPublicKey decodePublicKey(String value) throws InvalidKeyException {
        try {
            byte[] bytes = decode(value);
            PemObject pem = new PemObject("PUBLIC KEY", bytes);
            X509EncodedKeySpec spec = new X509EncodedKeySpec(pem.getContent());
            return (RSAPublicKey) KeyFactory.getInstance("RSA").generatePublic(spec);
        } catch (Exception e) {
            throw new InvalidKeyException("Invalid jwt public key", e);
        }
    }

    private static byte[] decode(String value) {
        try {
            return Base64.getUrlDecoder().decode(value);
        } catch (Exception e) {
            return Base64.getDecoder().decode(value);
        }
    }


}

File: iris-crypto/iris-crypto-key-rsa/src/main/kotlin/city/smartb/iris/crypto/rsa/RSAKeyPairDecoderBase64.kt
package city.smartb.iris.crypto.rsa

import java.security.InvalidKeyException
import java.security.KeyFactory
import java.security.KeyPair
import java.security.PrivateKey
import java.security.PublicKey
import java.security.interfaces.RSAPrivateKey
import java.security.interfaces.RSAPublicKey
import java.security.spec.PKCS8EncodedKeySpec
import java.security.spec.X509EncodedKeySpec
import java.util.Base64
import org.bouncycastle.util.io.pem.PemObject

object RSAKeyPairDecoderBase64 {

    @Throws(InvalidKeyException::class)
    fun decode(publicKey: String?, privateKey: String?): KeyPair {
        val pub: PublicKey = decodePublicKey(publicKey)
        val priv: PrivateKey = decodePrivateKey(privateKey)
        return KeyPair(pub, priv)
    }

    @Throws(InvalidKeyException::class)
    fun decodePrivateKey(value: String?): RSAPrivateKey {
        return try {
            val kf = KeyFactory.getInstance("RSA")
            val keySpecPKCS8 = PKCS8EncodedKeySpec(decode(value))
            kf.generatePrivate(keySpecPKCS8) as RSAPrivateKey
        } catch (e: Exception) {
            throw InvalidKeyException("Invalid private public key", e)
        }
    }

    @Throws(InvalidKeyException::class)
    fun decodePublicKey(value: String?): RSAPublicKey {
        return try {
            val bytes = decode(value)
            val pem = PemObject("PUBLIC KEY", bytes)
            val spec = X509EncodedKeySpec(pem.content)
            KeyFactory.getInstance("RSA").generatePublic(spec) as RSAPublicKey
        } catch (e: Exception) {
            throw InvalidKeyException("Invalid jwt public key", e)
        }
    }

    private fun decode(value: String?): ByteArray {
        return try {
            Base64.getUrlDecoder().decode(value)
        } catch (e: Exception) {
            Base64.getDecoder().decode(value)
        }
    }
}
